### PR TITLE
Explicitly complete scopes (do not use autoComplete: true)

### PR DESCRIPTION
**Pull request by ronaldbarendse**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description
In addition to adding an explicit way to rollback scopes (see PR https://github.com/umbraco/Umbraco-CMS/pull/12422), scopes shouldn't use `autoComplete: true` when mutating the database (or when this can't be guaranteed, e.g. because it's calling into user implemented methods). This is also mentioned as a remark on `CreateCoreScope()`:

https://github.com/umbraco/Umbraco-CMS/blob/60a5b19dc9fe24f5379a253e1f3fdcb9cb537dfc/src/Umbraco.Core/Scoping/ICoreScopeProvider.cs#L34-L35

~Because this is used in a lot of places, I've started the work and created this as a draft PR to have a way to share the progress... Maybe the best solution is to completely remove the `autoComplete` parameter, so we don't end up in the same situation in the future?~

@elit0451 updated more instances, after which some tests started failing because `CreateScope`/`CreateCoreScope` weren't mocked and returned `null` scopes. I've fixed this and continued with updating more instances where read/write locks were used (which uses `IDistributedLockingMechanism` and database access, potentially doing updates).

### CODE REVIEW

**Comment by ronaldbarendse** [comment]
To showcase the issue with using `autoComplete: true`, I've created some sample code that you can add to an existing Umbraco installation. An important thing to note is that notifications are published when the last (outer/parent) scope is disposed, but only if all scopes have completed (are not rolled back) and the database transaction is completed/committed.

```c#
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Core.Events;
using Umbraco.Cms.Core.Notifications;
using Umbraco.Cms.Core.Scoping;

public class ApplicationComposer : ComponentComposer<ApplicationComponent>
{
    public override void Compose(IUmbracoBuilder builder)
    {
        builder.AddNotificationHandler<ApplicationNotification, ApplicationNotificationHandler>();

        base.Compose(builder);
    }
}

public class ApplicationComponent : IComponent
{
    private readonly ICoreScopeProvider _scopeProvider;

    public ApplicationComponent(ICoreScopeProvider scopeProvider) => _scopeProvider = scopeProvider;

    public void Initialize()
    {
        // 1
        try
        {
            using (ICoreScope scope = _scopeProvider.CreateCoreScope(autoComplete: true))
            {
                scope.Notifications.Publish(new ApplicationNotification());

                throw new InvalidOperationException();
            }
        }
        catch (AggregateException ex) when (ex.InnerException is NotImplementedException)
        {
            // Probably not what you expect, because the InvalidOperationException was thrown first!
        }

        // 2
        try
        {
            using (ICoreScope scope = _scopeProvider.CreateCoreScope())
            {
                scope.Notifications.Publish(new ApplicationNotification());

                throw new InvalidOperationException();
            }
        }
        catch (InvalidOperationException)
        {
            // Makes sense, as the scope is not completed and therefore notifications are not published!
        }

        // 3
        try
        {
            using (ICoreScope scope = _scopeProvider.CreateCoreScope())
            {
                scope.Notifications.Publish(new ApplicationNotification());
                scope.Complete();
            }
        }
        catch (AggregateException ex) when (ex.InnerException is NotImplementedException)
        {
            // Again, makes sense!
        }

        // 4
        using (ICoreScope scope = _scopeProvider.CreateCoreScope())
        {
            scope.Notifications.Publish(new ApplicationNotification());

            // Scope is rolled back, so notifications are not published and no exception is thrown
        }

        // 5
        using (ICoreScope parentScope = _scopeProvider.CreateCoreScope())
        {
            using (ICoreScope childScope = _scopeProvider.CreateCoreScope())
            {
                childScope.Notifications.Publish(new ApplicationNotification());
                childScope.Complete();
            }

            // Parent scope is rolled back, so notifications are not published and no exception is thrown
        }

        // 6
        using (ICoreScope parentScope = _scopeProvider.CreateCoreScope())
        {
            parentScope.Notifications.Publish(new ApplicationNotification());

            using (ICoreScope childScope = _scopeProvider.CreateCoreScope())
            {
                // Child scope is not completed and therefore causes the parent scope to also rollback
            }

            parentScope.Complete();

            // Child scope is rolled back, so notifications are not published and no exception is thrown
        }

        // 7
        {
            ICoreScope scope = _scopeProvider.CreateCoreScope();
            scope.Notifications.Publish(new ApplicationNotification());
            scope.Complete();

            try
            {
                scope.Dispose();
            }
            catch (AggregateException ex) when (ex.InnerException is NotImplementedException)
            {
                // Just to show the notifications are published when disposing, not completing, the scope!
            }
        }
    }

    public void Terminate()
    { }
}

public class ApplicationNotification : INotification
{ }

public class ApplicationNotificationHandler : INotificationHandler<ApplicationNotification>
{
    public void Handle(ApplicationNotification notification) => throw new NotImplementedException();
}
```

**Comment by ronaldbarendse** [comment]
> **Warning** Carefully review all scopes that don't have `autoComplete: true` anymore now call `scope.Complete()` in _all_ code paths (also before throwing exceptions or returning early). If this isn't done, the scope will roll back and cause hard to debug issues (see https://github.com/umbraco/Umbraco-CMS/pull/12293#discussion_r864650719)!

> **Note** I've simplified usings that still use `autoComplete: true` to not capture the returned scope, as you shouldn't publish notifications on auto-completing scopes or take read/write locks, so any remaining instances should only use `using (ScopeProvider.CreateCoreScope(autoComplete: true)) {}`.

File: src/Umbraco.Core/Services/MemberService.cs [hunk 1]
```diff
@@ -76,10 +76,14 @@ public int GetCount(MemberCountType countType)
                     query = Query<IMember>()?.Where(x => x.IsApproved == true);
                     break;
                 default:
+                    scope.Complete();
```
[the remark this answers is not in the document]
Comment (ronaldbarendse) [reply src/Umbraco.Core/Services/MemberService.cs:79]: Or we actually don't complete the scope anymore and rollback because of this exception?

File: src/Umbraco.Core/Services/MemberTypeService.cs [hunk 2]
```diff
@@ -73,14 +73,15 @@ public class MemberTypeService : ContentTypeServiceBase<IMemberTypeRepository, I
...
-        using (ICoreScope scope = ScopeProvider.CreateCoreScope(autoComplete: true))
+        using (ICoreScope scope = ScopeProvider.CreateCoreScope())
         {
             scope.ReadLock(ReadLockIds);
 
             using (IEnumerator<IMemberType> e = _memberTypeRepository.GetMany(new int[0]).GetEnumerator())
             {
                 if (e.MoveNext() == false)
                 {
+                    scope.Complete();
```
Comment (ronaldbarendse) [inline src/Umbraco.Core/Services/MemberTypeService.cs:84]: Again: we might want to rewrite or remove the `Complete()` call before throwing an exception.
